WalletWasabi.Gui/Tabs/EncryptionManager/VerifyMessageViewModel.cs

```diff
@@ -0,0 +1,114 @@
+using WalletWasabi.Gui.Tabs.WalletManager;
+using WalletWasabi.Gui.ViewModels;
+using WalletWasabi.Helpers;
+using WalletWasabi.KeyManagement;
+
+namespace WalletWasabi.Gui.Tabs.EncryptionManager
+{
+	internal class VerifyMessageViewModel : CategoryViewModel
+	{
+		private string _message;
+		private string _address;
+		private string _password;
+		private string _signature;
+		private string _warningMessage;
+
+		public string Message
+		{
+			get => _message;
+			set => this.RaiseAndSetIfChanged(ref _message, value);
+		}
+
+		public string Address
+		{
+			get => _address;
+			set => this.RaiseAndSetIfChanged(ref _address, value);
+		}
+
+		public string Password
+		{
+			get => _password;
+			set => this.RaiseAndSetIfChanged(ref _password, value);
+		}
+
+		public string Signature
+		{
+			get => _signature;
+			set => this.RaiseAndSetIfChanged(ref _signature, value);
+		}
+
+		public string WarningMessage
+		{
+			get => _warningMessage;
+			set => this.RaiseAndSetIfChanged(ref _warningMessage, value);
+		}
+
+		public ReactiveCommand SignCommand { get; }
+		public ReactiveCommand VerifyCommand { get; }
+		public EncryptionManagerViewModel Owner { get; }
+
+		public VerifyMessageViewModel(EncryptionManagerViewModel owner) : base("Verify Message")
+		{
+			Owner = owner;
+
+			var canSign = this.WhenAnyValue(x => x.Message, x => x.Address,
+				(message, address) =>
+					!string.IsNullOrEmpty(message) &&
+					!string.IsNullOrEmpty(address));
+
+			var canVerify = this.WhenAnyValue(x => x.Message, x => x.Address, x => x.Signature,
+				(message, address, sign) =>
+					!string.IsNullOrEmpty(message) &&
+					!string.IsNullOrEmpty(address) &&
+					!string.IsNullOrEmpty(sign));
+
+			SignCommand = ReactiveCommand.Create(
+				() =>
+				{
+					Signature = SignMessage(Address, Message, Password);
+				},
+				canSign
+			);
+			SignCommand.ThrownExceptions.Subscribe(ex =>
+			{
+				WarningMessage = ex.Message;
+			});
+
+			VerifyCommand = ReactiveCommand.Create(
+				() =>
+				{
+					var verified = VerifyMessage(Address, Message, Signature);
+					if (!verified) throw new InvalidOperationException("Invalid signature");
+					WarningMessage = "Good";
+				}
+				, canVerify
+			);
+			VerifyCommand.ThrownExceptions.Subscribe(ex =>
+			{
+				WarningMessage = ex.Message;
+			});
+		}
+
+		private static string SignMessage(string address, string message, string password)
+		{
+			//https://programmingblockchain.gitbook.io/programmingblockchain/bitcoin_transfer/proof_of_ownership_as_an_authentication_method
+			password = Guard.Correct(password);
+			BitcoinSecret bitcoinPrivateKey = Global.WalletService.KeyManager.EncryptedSecret.GetSecret(password);
+			string signature = bitcoinPrivateKey.PrivateKey.SignMessage(message);
+			return signature;
+		}
+
+		private static bool VerifyMessage(string address, string message, string signature)
+		{
+			BitcoinPubKeyAddress addr = new BitcoinPubKeyAddress(address, Global.Network);
```

Comment:
Same that previous comment. `BitcoinPubKeyAddress` is an address for what is call Pay-To-PubKey-Hash (P2PKH) and you can recognize the easily because they start with 1. Wasabi wallet doesn't support address of that type. We only support a new type of bitcoin address called Pay-To-Witness-PubKey-Hash (P2WPKH) that are represented by the class `BitcoinWitPubKeyAddress` and start with 'bc1'